src/vcpkg/paragraphs.cpp

```diff
@@ -20,139 +20,126 @@ static std::atomic<uint64_t> g_load_ports_stats(0);
-            }
-
-            target.push_back('\n');
-            target.append(expected_types_component.extract_data());
-        }
+        return std::string();
     }
 
-    std::string ParseControlErrorInfo::to_string() const
+    std::string ParagraphParser::optional_field(StringLiteral fieldname)
     {
-        std::string result;
-        to_string(result);
-        return result;
+        TextRowCol ignore;
+        return optional_field(fieldname, ignore);
     }
 
-    LocalizedString ToLocalizedString_t::operator()(std::unique_ptr<ParseControlErrorInfo> p) const
+    std::string ParagraphParser::required_field(StringLiteral fieldname)
     {
-        return LocalizedString::from_raw(p->to_string());
+        auto maybe_field = remove_field(&fields, fieldname);
+        if (const auto field = maybe_field.get())
+        {
+            return std::move(field->first);
+        }
+
+        errors.emplace_back(LocalizedString::from_raw(origin)
+                                .append_raw(": ")
+                                .append_raw(ErrorPrefix)
+                                .append(msgMissingRequiredField2, msg::json_field = fieldname));
+        return std::string();
     }
 
-    std::unique_ptr<ParseControlErrorInfo> ParseControlErrorInfo::from_error(StringView name, LocalizedString&& ls)
+    void ParagraphParser::add_error(TextRowCol position, msg::MessageT<> error_content)
     {
-        auto error_info = std::make_unique<ParseControlErrorInfo>();
-        error_info->name.assign(name.data(), name.size());
-        error_info->error = std::move(ls);
-        return error_info;
+        errors.emplace_back(LocalizedString::from_raw(origin)
+                                .append_raw(fmt::format("{}:{}: ", position.row, position.column))
+                                .append_raw(ErrorPrefix)
+                                .append(error_content));
     }
+} // namespace vcpkg
 
-    static Optional<std::pair<std::string, TextRowCol>> remove_field(Paragraph* fields, StringView fieldname)
+namespace
+{
+    void append_errors(LocalizedString& result, const std::vector<LocalizedString>& errors)
     {
-        auto it = fields->find(fieldname.to_string());
-        if (it == fields->end())
+        auto error = errors.begin();
+        const auto last = errors.end();
+        for (;;)
         {
-            return nullopt;
-        }
+            result.append(*error);
+            if (++error == last)
+            {
+                return;
+            }
 
-        auto value = std::move(it->second);
-        fields->erase(it);
-        return value;
+            result.append_raw('\n');
+        }
     }
 
-    void ParagraphParser::required_field(StringLiteral fieldname, std::pair<std::string&, TextRowCol&> out)
+    void append_field_errors(LocalizedString& result, const std::string& origin, const Paragraph& fields)
```

Comment:
```suggestion
    void append_field_errors(LocalizedString& result, StringView origin, const Paragraph& fields)
```

Will hit StringView overload of `append_raw()` anyways